src/drivers/intel/ssp/ssp.c

```diff
@@ -74,49 +74,74 @@ static void ssp_empty_tx_fifo(struct dai *dai)
 		ssp_write(dai, SSSR, sssr);
 }
 
-/* empty SSP receive FIFO */
-static void ssp_empty_rx_fifo(struct dai *dai)
+static void ssp_empty_rx_fifo_on_start(struct dai *dai)
+{
+	uint32_t retry = SSP_RX_FLUSH_RETRY_MAX;
+	uint32_t i, sssr;
+
+	sssr = ssp_read(dai, SSSR);
+
+	if (sssr & SSSR_ROR) {
+		/* The RX FIFO is in overflow condition, empty it */
+		for (i = 0; i < 32; i++)
+			ssp_read(dai, SSDR);
```

Comment:
Is this safe. The maximum FIFO length is 16 right?

Reply:
It appears to be 32 deep on TGL at least (upx-i11).
```diff
diff --git a/src/drivers/intel/ssp/ssp.c b/src/drivers/intel/ssp/ssp.c
index a75f8752ed0a..cb0b076ebc37 100644
--- a/src/drivers/intel/ssp/ssp.c
+++ b/src/drivers/intel/ssp/ssp.c
@@ -81,7 +81,15 @@ static void ssp_empty_rx_fifo_on_start(struct dai *dai)
 
 	sssr = ssp_read(dai, SSSR);
 
+	dai_info(dai, "ssp_empty_rx_fifo_on_start(): initial SSRSA = 0x%08x",
+		 ssp_read(dai, SSRSA));
+	dai_info(dai, "ssp_empty_rx_fifo_on_start(): initial SSSR = 0x%08x",
+		 ssp_read(dai, SSSR));
+	dai_info(dai, "ssp_empty_rx_fifo_on_start(): initial entries = %u",
+		 (uint32_t)SSCR3_RFL_VAL(ssp_read(dai, SSCR3)));
+
 	if (sssr & SSSR_ROR) {
+		dai_info(dai, "ssp_empty_rx_fifo_on_start(): yanking FIFO");
 		/* The RX FIFO is in overflow condition, empty it */
 		for (i = 0; i < 32; i++)
 			ssp_read(dai, SSDR);
@@ -90,17 +98,30 @@ static void ssp_empty_rx_fifo_on_start(struct dai *dai)
 		ssp_update_bits(dai, SSSR, SSSR_ROR, SSSR_ROR);
 		/* Re-read the SSSR register */
 		sssr = ssp_read(dai, SSSR);
+		dai_info(dai, "ssp_empty_rx_fifo_on_start(): entries after yank = %u",
+			 (uint32_t)SSCR3_RFL_VAL(ssp_read(dai, SSCR3)));
 	}
 
+	dai_info(dai, "ssp_empty_rx_fifo_on_start(): before loop: SSSR = 0x%08x", sssr);
 	while ((sssr & SSSR_RNE) && retry--) {
 		uint32_t entries = SSCR3_RFL_VAL(ssp_read(dai, SSCR3));
 
+		dai_info(dai,
+			 "ssp_empty_rx_fifo_on_start(): retry round: %d: entries: %u",
+			 SSP_RX_FLUSH_RETRY_MAX - retry, entries);
 		/* Empty the RX FIFO (the DMA is not running at this point) */
 		for (i = 0; i < entries + 1; i++)
 			ssp_read(dai, SSDR);
 
 		sssr = ssp_read(dai, SSSR);
+		dai_info(dai,
+			 "ssp_empty_rx_fifo_on_start(): end of retry round: %d: ssr: 0x%08x",
+			 SSP_RX_FLUSH_RETRY_MAX - retry, sssr);
 	}
+
+	sssr = ssp_read(dai, SSSR);
+	dai_info(dai,
+		 "ssp_empty_rx_fifo_on_start(): after all of this: SSSR = 0x%08x", sssr);
 }
 
 static void ssp_empty_rx_fifo_on_stop(struct dai *dai)
```
With this PR (FIFO depth is 32):
```
/drivers/intel/ssp/ssp.c:84   INFO ssp_empty_rx_fifo_on_start(): initial SSRSA = 0x00000001
/drivers/intel/ssp/ssp.c:86   INFO ssp_empty_rx_fifo_on_start(): initial SSSR = 0x000000dc
/drivers/intel/ssp/ssp.c:88   INFO ssp_empty_rx_fifo_on_start(): initial entries = 0
/drivers/intel/ssp/ssp.c:92   INFO ssp_empty_rx_fifo_on_start(): yanking FIFO
/drivers/intel/ssp/ssp.c:101  INFO ssp_empty_rx_fifo_on_start(): entries after yank = 0
/drivers/intel/ssp/ssp.c:105  INFO ssp_empty_rx_fifo_on_start(): before loop: SSSR = 0x00000014
/drivers/intel/ssp/ssp.c:123  INFO ssp_empty_rx_fifo_on_start(): after all of this: SSSR = 0x00000014
```
Using `SSP_FIFO_DEPTH`, which is defined to 16 (so FIFO depth of 16):
```
/drivers/intel/ssp/ssp.c:84   INFO ssp_empty_rx_fifo_on_start(): initial SSRSA = 0x00000001
/drivers/intel/ssp/ssp.c:86   INFO ssp_empty_rx_fifo_on_start(): initial SSSR = 0x000000dc
/drivers/intel/ssp/ssp.c:88   INFO ssp_empty_rx_fifo_on_start(): initial entries = 0
/drivers/intel/ssp/ssp.c:92   INFO ssp_empty_rx_fifo_on_start(): yanking FIFO
/drivers/intel/ssp/ssp.c:101  INFO ssp_empty_rx_fifo_on_start(): entries after yank = 16
/drivers/intel/ssp/ssp.c:105  INFO ssp_empty_rx_fifo_on_start(): before loop: SSSR = 0x0000005c
/drivers/intel/ssp/ssp.c:109  INFO ssp_empty_rx_fifo_on_start(): retry round: 1: entries: 16
/drivers/intel/ssp/ssp.c:117  INFO ssp_empty_rx_fifo_on_start(): end of retry round: 1: ssr: 0x00000014
/drivers/intel/ssp/ssp.c:123  INFO ssp_empty_rx_fifo_on_start(): after all of this: SSSR = 0x0000001c
```
In a later case (assuming depth of 16 for the FIFO) after the yanking of the FIFO we still have 16 entries left (which is an odd result as the number of entries in FIFO is +1, so 17?).
It looks awfully like that the FIFO is 32 deep by this experiment. I'm fine to use the two step yanking assuming a 16 word deep FIFO, at the end it should not matter much.